### PR TITLE
PROTON-2481: [cpp] Improve constructor syntax for maps

### PR DESCRIPTION
**Pull request by DreamPearl**
[PROTON-2481](https://issues.apache.org/jira/browse/PROTON-2481)
- Add a copy constructor to copy from std::map
- Add an initializer_list constructor

### CODE REVIEW

File: cpp/include/proton/map.hpp [hunk 1]
```diff
@@ -26,6 +26,8 @@
 #include "./internal/pn_unique_ptr.hpp"
 
 #include <cstddef>
+#include <initializer_list>
+#include <map>
```
Comment (astitcher) [inline cpp/include/proton/map.hpp:30]: I'm not sure this is correct given my comment below - you may need pair though as that is used in the initializer_list constructor

File: cpp/include/proton/map.hpp [hunk 2]
```diff
@@ -79,6 +81,12 @@ class PN_CPP_CLASS_EXTERN map {
     /// Copy a map.
     PN_CPP_EXTERN map& operator=(const map&);
 
+    /// Copy a std::map.
+    PN_CPP_EXTERN map(const std::map<K, T>&);
```
Comment (astitcher) [inline cpp/include/proton/map.hpp:85]: I think this is too specific: The assigment operator uses a template to take any argument that is assignable to a proton::value: as a map - this constructor should do the same.
The assignment operator uses enable_if to ensure that it is the template type will work - I'm not sure you can use it in exactly the same way for a constructor. Actually I think we should replace the custom enable_if with std::enable_if now that we can use C++ 11

**Review by astitcher — CHANGES_REQUESTED** [review]

**Comment by jiridanek** [comment]
There aren't any usages for the new constructors. I'd like to see some test that calls them.

**Comment by astitcher** [comment]
Merged

**Pull request closed without merging**